### DISCUSSION TITLE
Auth.js not resolve correct Host (Origin) on redirectUrl, SignIn Page

### DISCUSSION THREAD

**Discussion by LordLafayette** [nextauthjs/next-auth#11331] [Help]
When incoming request come with different Domain, eg. "fuba.com", Auth.js not reslove with it domain, but will be localhost:3000 instead

**On Develop**

Using command `yarn dev`

Server listening at: http://localhost:3000

Map domain "fubar.com" with IP 127.0.0.1

Request URL > https://fubar.com:3000/api/auth/signin

### Problem

**On Sign-In Page  (/api/auth/signin)**

When inspect to the Form, The URL that from will submit sign-in is: http://localhost:3000/api/auth/som-provider, instead of incoming domain (fubar.com) .

**At Redeem access_token after get authorizeCode**

When signin('provier') call, auth.js to redirect browser agent to AuthorizeEndpoint with redirect_url **( In this step the auth.js reslove domain correctly with "fubar.com")**.

**But** when Auth.js try to redeem token from AuthorizeCode, it send "local server host (localhost:3000)" at redirect_url to TokenEndpoint, That cause OP to invalid request.

***This problem aslo appear on deployment with self host (yarn build & start) and Docker*** 

### Solution

The issue was solved when set `AUTH_URL` = production URL or any host name, But this setting will have limit when try to use application with multiple domain name 

BTW, as you have mentioned in Doc (https://authjs.dev/getting-started/deployment#auth_url)

> This environment variable is mostly unnecessary with v5 as the host is inferred from the request headers

Why it not resolve HOST from request header  `host`

```
{
 nextUrl: {
  href: 'http://localhost:3000/next.svg',
  origin: 'http://localhost:3000',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost:3000',
  hostname: 'localhost',
  port: '3000',
  pathname: '/next.svg',
  search: '',
  searchParams: URLSearchParams {  },
  hash: ''
},
headers: {
  accept: 'image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8',
  accept-encoding: 'gzip, deflate',
  accept-language: 'th,en-US;q=0.9,en;q=0.8,en-GB;q=0.7',
  connection: 'keep-alive',
  referer: 'http://wakanda.com:3000/',
  host: 'wakanda.com:3000',
  user-agent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0',
  x-forwarded-for: '::ffff:127.0.0.1',
  x-forwarded-host: 'wakanda.com:3000',
  x-forwarded-port: '3000',
  x-forwarded-proto: 'http'
}
}
```

Thankyou.

```
Windows 11
Nojde.js : LTS (20)
Next.js: 14.2.4
next-auth: 5.0.0-beta.19
```
### On Self Host `yarn build & yarn start` and  On Dev `yarn dev`

![image](https://github.com/nextauthjs/next-auth/assets/30994730/7e17e8d0-e98f-4cd1-ad72-705311e867af)

### On Docker

![Screenshot 2024-07-05 030950](https://github.com/nextauthjs/next-auth/assets/30994730/54dc97e0-e1e1-4478-9bf9-47699d452cdb)

**Comment by LordLafayette** [comment]
Additionally

I have follow code by this example: https://authjs.dev/getting-started/installation

This is provider config


```
import type { NextAuthConfig } from 'next-auth';
import azure from 'next-auth/providers/microsoft-entra-id'

const nextAuthConfig = {
  providers: [
    azure({
        clientId:'id',
        clientSecret:'secret',        
    })
  ],
  secret:'auth.js secret'
} satisfies NextAuthConfig;

export default nextAuthConfig;
```

**Comment by LordLafayette** [comment]
I have noticed that why it resolve Host correct on redirect to Authorize Endpoint , but wrong at Redeem access_token step after receive authorize_code

_PS. signIn() method call from route handler, when request not authenticated yet, The middleware detect and redirect to route "/auth/azure-sign-in"_


`/app/auth/azure-sign-in/route.js`

```

import { signIn } from 'src/auth.js';

/* -------------------------------------------------------------------------- */

export async function GET(request: Request) {
  const { searchParams } = new URL(request.url);

  const redirectTo= searchParams.get('returnTo') ?? '/dashboard';

  await signIn('local-oidc',{redirectTo});
}

```

`middleware.ts`

```
export const middleware = auth((req) => {
  const isDashboard = req.nextUrl.pathname.startsWith('/dashboard');
  if (isDashboard && !req.auth) {
    const urlSearchParams = new URLSearchParams();
    urlSearchParams.set('returnTo', `${req.nextUrl.pathname}${req.nextUrl.search}`);

    const loginPath = `/auth/azure-sign-in?${urlSearchParams.toString()}`;
    const loginUrl = new URL(loginPath, req.nextUrl.origin);

    return NextResponse.redirect(loginUrl);
  }
});
```

**Reply by Feeco-Li** [reply]
is this a solution?

next-auth is having so many open issue like this. 

I'm facing same problem.

**Reply by shawnmclean** [reply]
I'm testing the website on a 192.xx over LAN and having this same issue.

I thought it was looking at the actual hostnames of the origin?
